File: src/scope.ts
/**
 * Scope — request context, user context, and tags.
 * Uses AsyncLocalStorage for per-request isolation in Node.js.
 * Falls back to global state in edge runtime.
 *
 * ── Two-layer redaction model ────────────────────────────────────────────
 *
 * This module runs an always-on baseline scrub at `setRequestContext()`
 * time. It (a) redacts headers whose name matches a sensitive pattern
 * (`token` / `key` / `auth` / …), (b) wholesales sensitive body fields
 * (matched against the shared `SENSITIVE_KEYS` set imported from
 * `redact/keys.ts`), (c) trims oversized strings and arrays, and
 * (d) hard-redacts forwarded-IP headers.
 *
 * The full `redact/` module is OPT-IN (`init({ redact: true })`) and
 * complements this baseline: regex-based content scrubs (email, JWT,
 * Stripe / AWS / GitHub keys, Luhn-validated card numbers), allowlists,
 * hash mode, etc., over the whole event at send time.
 *
 * Both layers share `SENSITIVE_KEYS` so the field-name list cannot drift
 * between them.
 */

import { SENSITIVE_KEYS } from "./redact/keys.js"

let asyncStorage: any = null
try {
  const { AsyncLocalStorage } = require("node:async_hooks")
  asyncStorage = new AsyncLocalStorage()
} catch {
  // Edge runtime — fallback to global
}

interface Scope {
  user?: { id?: string; role?: string }
  tags?: Record<string, string>
  requestContext?: {
    method: string
    url: string
    headers?: Record<string, string>
    query?: Record<string, string>
    body?: unknown
    ip?: string
  }
}

let globalScope: Scope = {}

// Pattern-based: redact any header whose NAME contains one of these
// substrings. Complements `SENSITIVE_KEYS` (which is exact-match only).
const REDACT_HEADER_PATTERNS = ["token", "key", "secret", "auth", "credential", "password", "cookie", "session"]

function shouldRedactHeader(name: string): boolean {
  const lower = name.toLowerCase()
  return REDACT_HEADER_PATTERNS.some((p) => lower.includes(p))
}

function redactBody(body: unknown): unknown {
  if (body === null || body === undefined) return body
  if (typeof body === "string") {
    return body.length > 1024 ? body.slice(0, 1024) + "...[truncated]" : body
  }
  if (typeof body !== "object") return body
  if (Array.isArray(body)) return body.slice(0, 20)

  const safe: Record<string, unknown> = {}
  const obj = body as Record<string, unknown>
  for (const [k, v] of Object.entries(obj)) {
    if (SENSITIVE_KEYS.has(k.toLowerCase())) {
      safe[k] = "[REDACTED]"
    } else if (typeof v === "string" && v.length > 500) {
      safe[k] = v.slice(0, 500) + "...[truncated]"
    } else {
      safe[k] = v
    }
  }
  return safe
}

function getScope(): Scope {
  if (asyncStorage) {
    return asyncStorage.getStore() ?? globalScope
  }
  return globalScope
}

// ── Public API ──────────────────────────────────────────────────────────────

export function setUser(user: { id?: string; email?: string; role?: string }): void {
  // Strip email by default (PII) — only keep id + role
  const safe = { id: user.id, role: user.role }
  const scope = getScope()
  scope.user = safe
}

export function setTag(key: string, value: string): void {
  const scope = getScope()
  if (!scope.tags) scope.tags = {}
  scope.tags[key] = value
}

export function setRequestContext(ctx: {
  method: string
  url: string
  headers?: Record<string, string>
  query?: Record<string, string>
  body?: unknown
  ip?: string
}): void {
  const scope = getScope()

  // Redact sensitive headers (pattern-based)
  const safeHeaders: Record<string, string> = {}
  if (ctx.headers) {
    for (const [k, v] of Object.entries(ctx.headers)) {
      safeHeaders[k] = shouldRedactHeader(k) ? "[REDACTED]" : v
    }
    // Also redact IP-related headers
    for (const h of ["x-forwarded-for", "x-real-ip"]) {
      if (safeHeaders[h]) safeHeaders[h] = "[REDACTED]"
    }
  }

  scope.requestContext = {
    method: ctx.method,
    url: ctx.url,
    headers: Object.keys(safeHeaders).length > 0 ? safeHeaders : undefined,
    query: ctx.query,
    body: redactBody(ctx.body),
    // IP omitted by default (PII/GDPR) — only sent if explicitly set
  }
}

export function getUser(): Scope["user"] {
  return getScope().user
}

export function getTags(): Scope["tags"] {
  return getScope().tags
}

export function getRequestContext(): Scope["requestContext"] {
  return getScope().requestContext
}

/**
 * Run a function with an isolated scope (for per-request isolation).
 * Use in middleware: runWithScope(() => handleRequest(req, res))
 */
export function runWithScope<T>(fn: () => T): T {
  if (asyncStorage) return asyncStorage.run({}, fn);
  const prev = globalScope;
  globalScope = {};
  try { return fn(); }
  finally { globalScope = prev; }
}

File: go/scope.go
// Per-request scope via context.Context (Go's idiomatic equivalent of
// Python contextvars / Node AsyncLocalStorage).
//
// SetUser / SetTag / SetRequestContext mutate either:
//   - the scope stored in the current context.Context (preferred), or
//   - a process-global fallback scope (for simple scripts with no ctx).
//
// Middleware should always call WithScope(ctx, fn) — that gives each
// request an isolated scope instance and prevents leakage between
// concurrent requests.
package capture

import (
	"context"
	"strings"
	"sync"
)

type scopeData struct {
	mu       sync.RWMutex
	user     *User
	tags     map[string]string
	request  *RequestContext
}

// Context key type — unexported to avoid collision with other packages.
type scopeKey struct{}

var globalScope = &scopeData{}

func currentScope(ctx context.Context) *scopeData {
	if ctx != nil {
		if s, ok := ctx.Value(scopeKey{}).(*scopeData); ok {
			return s
		}
	}
	return globalScope
}

// WithScope returns a new context that owns a fresh, isolated scope.
// Use in middleware: ctx = WithScope(ctx); defer ctx.Done().
func WithScope(parent context.Context) context.Context {
	if parent == nil {
		parent = context.Background()
	}
	return context.WithValue(parent, scopeKey{}, &scopeData{})
}

// ── User/tag/request setters ────────────────────────────────────────

// SetUser attaches user context. Email is always stripped for privacy
// — only ID + Role survive. Pass ctx=nil to write to the global scope.
func SetUser(ctx context.Context, user User) {
	safe := &User{ID: user.ID, Role: user.Role}
	s := currentScope(ctx)
	s.mu.Lock()
	s.user = safe
	s.mu.Unlock()
}

// SetTag accumulates custom tags on the current scope.
func SetTag(ctx context.Context, key, value string) {
	s := currentScope(ctx)
	s.mu.Lock()
	if s.tags == nil {
		s.tags = make(map[string]string)
	}
	s.tags[key] = value
	s.mu.Unlock()
}

// Header redaction — matches Python/Node SDK patterns.
var redactHeaderPatterns = []string{
	"token", "key", "secret", "auth",
	"credential", "password", "cookie", "session",
}

func shouldRedactHeader(name string) bool {
	lower := strings.ToLower(name)
	for _, p := range redactHeaderPatterns {
		if strings.Contains(lower, p) {
			return true
		}
	}
	return false
}

// RedactHeaders returns a copy of the header map with sensitive values
// replaced by ``[REDACTED]``. Exposed publicly so framework middleware
// can scrub headers before storing them on the scope.
func RedactHeaders(in map[string]string) map[string]string {
	if in == nil {
		return nil
	}
	out := make(map[string]string, len(in))
	for k, v := range in {
		if shouldRedactHeader(k) {
			out[k] = "[REDACTED]"
		} else {
			out[k] = v
		}
	}
	return out
}

// Body fields to always replace with [REDACTED]. Matches the Node
// REDACT_BODY_FIELDS set.
var redactBodyFields = map[string]struct{}{
	"password": {}, "passwd": {}, "pass": {}, "secret": {}, "token": {},
	"api_key": {}, "apiKey": {}, "access_token": {}, "accessToken": {},
	"refresh_token": {}, "refreshToken": {}, "credit_card": {}, "creditCard": {},
	"card_number": {}, "cardNumber": {}, "cvv": {}, "cvc": {}, "ssn": {},
	"social_security": {}, "authorization": {},
}

func isRedactedField(k string) bool {
	if _, ok := redactBodyFields[k]; ok {
		return true
	}
	_, ok := redactBodyFields[strings.ToLower(k)]
	return ok
}

// RedactBody returns a copy of ``body`` with sensitive keys scrubbed.
// Exposed publicly so middleware can redact payloads before attaching
// them to the scope.
func RedactBody(body interface{}) interface{} {
	switch v := body.(type) {
	case nil:
		return nil
	case string:
		const maxLen = 1024
		if len(v) > maxLen {
			return v[:maxLen] + "...[truncated]"
		}
		return v
	case map[string]interface{}:
		safe := make(map[string]interface{}, len(v))
		for k, val := range v {
			if isRedactedField(k) {
				safe[k] = "[REDACTED]"
				continue
			}
			if s, ok := val.(string); ok && len(s) > 500 {
				safe[k] = s[:500] + "...[truncated]"
				continue
			}
			safe[k] = val
		}
		return safe
	case []interface{}:
		const maxItems = 20
		if len(v) > maxItems {
			return v[:maxItems]
		}
		return v
	default:
		return v
	}
}

// SetRequestContext attaches request metadata. Sensitive headers and
// body fields are redacted before storage. IP is included only when
// the caller explicitly sets it (GDPR-friendly default).
func SetRequestContext(ctx context.Context, rc RequestContext) {
	safe := RequestContext{
		Method: rc.Method,
		URL:    rc.URL,
		Query:  rc.Query,
	}
	if len(rc.Headers) > 0 {
		safeHeaders := make(map[string]string, len(rc.Headers))
		for k, v := range rc.Headers {
			if shouldRedactHeader(k) {
				safeHeaders[k] = "[REDACTED]"
				continue
			}
			safeHeaders[k] = v
		}
		// IP-bearing headers scrubbed too.
		for h := range safeHeaders {
			if lower := strings.ToLower(h); lower == "x-forwarded-for" || lower == "x-real-ip" {
				safeHeaders[h] = "[REDACTED]"
			}
		}
		safe.Headers = safeHeaders
	}
	if rc.Body != nil {
		safe.Body = RedactBody(rc.Body)
	}
	if rc.IP != "" {
		safe.IP = rc.IP
	}

	s := currentScope(ctx)
	s.mu.Lock()
	s.request = &safe
	s.mu.Unlock()
}

// GetUser returns the current scope's user, or nil.
func GetUser(ctx context.Context) *User {
	s := currentScope(ctx)
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.user
}

// GetTags returns a copy of the current scope's tags.
func GetTags(ctx context.Context) map[string]string {
	s := currentScope(ctx)
	s.mu.RLock()
	defer s.mu.RUnlock()
	if len(s.tags) == 0 {
		return nil
	}
	out := make(map[string]string, len(s.tags))
	for k, v := range s.tags {
		out[k] = v
	}
	return out
}

// GetRequestContext returns the current scope's request context, or nil.
func GetRequestContext(ctx context.Context) *RequestContext {
	s := currentScope(ctx)
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.request
}

// ClearScope resets the active scope — primarily for tests.
func ClearScope(ctx context.Context) {
	s := currentScope(ctx)
	s.mu.Lock()
	s.user = nil
	s.tags = nil
	s.request = nil
	s.mu.Unlock()
}
